src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -531,5 +542,707 @@ public void Add(NamespaceOrTypeSymbol symbol)
+                                                               GetExternAliases(declaringSymbol, declarationSyntax)),
+                                                null);
+                }
+
+                return _lazyImports;
+            }
+
+            internal void Complete(SourceNamespaceSymbol declaringSymbol, SyntaxReference declarationSyntax, CancellationToken cancellationToken)
+            {
+                var externAliasesAndDiagnostics = _lazyExternAliases ?? GetExternAliasesAndDiagnostics(declaringSymbol, (CSharpSyntaxNode)declarationSyntax.GetSyntax());
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var usingsAndDiagnostics = _lazyUsings ?? GetUsingsAndDiagnostics(declaringSymbol, (CSharpSyntaxNode)declarationSyntax.GetSyntax(), basesBeingResolved: null);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                while (true)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var incompletePart = _state.NextIncompletePart;
+                    switch (incompletePart)
+                    {
+                        case CompletionPart.StartValidatingImports:
+                            {
+                                if (_state.NotePartComplete(CompletionPart.StartValidatingImports))
+                                {
+                                    Validate(declaringSymbol, externAliasesAndDiagnostics, usingsAndDiagnostics);
+                                    _state.NotePartComplete(CompletionPart.FinishValidatingImports);
+                                }
+                            }
+                            break;
+
+                        case CompletionPart.FinishValidatingImports:
+                            // some other thread has started validating imports (otherwise we would be in the case above) so
+                            // we just wait for it to both finish and report the diagnostics.
+                            Debug.Assert(_state.HasComplete(CompletionPart.StartValidatingImports));
+                            _state.SpinWaitComplete(CompletionPart.FinishValidatingImports, cancellationToken);
+                            break;
+
+                        case CompletionPart.None:
+                            return;
+
+                        default:
+                            // any other values are completion parts intended for other kinds of symbols
+                            _state.NotePartComplete(CompletionPart.All & ~CompletionPart.ImportsAll);
+                            break;
+                    }
+
+                    _state.SpinWaitComplete(incompletePart, cancellationToken);
+                }
+            }
+
+            private static void Validate(SourceNamespaceSymbol declaringSymbol, ExternAliasesAndDiagnostics externAliasesAndDiagnostics, UsingsAndDiagnostics usingsAndDiagnostics)
```

Comment:
>Validate [](start = 32, length = 8)

This is adapted flavor of Imports.Validate. What used to be members now accessed through parameters. Also some extra handling of usings from options (null checks for syntax) is removed.